### PR TITLE
Async password verifier

### LINKED ISSUE #2218: Digest passwords on thread pool
I am using the following call to authenticate the user:

    User.authenticate(
        username: username,
        password: password,
        using: passwordVerifier,
        on: $0)

This runs `guard try verifier.verify(basic.password, created: user.basicPassword) else {` after fetching the user's encrypted password. If the verifier is slow, e.g. `bcrypt`-based, this verification might take up to a few 100 milliseconds (but even 50-100 would already be bad). During that time, the event loop can not process any other events, meaning that any other operations on the same loop (e.g. other requests, DB calls, etc.) get delayed by this much. In particular, authenticating one user's request might noticeably delay another user's request just by the virtue of being scheduled on the same event loop.

To avoid this, I'd suggest running `verifier.verify` off the event loop (possibly on a concurrent dispatch queue) and, once that is done, fulfilling a subsequent promise back on the event loop.

What do you think? Also, which way of moving this work off the event loop would you prefer? In fact, should Vapor offer a "standard"/"best practice" method of doing this?

### PR DESCRIPTION
**Pull request by madsodgaard**
This PR adds `AsyncPasswordVerifier` to allow slow methods like `BCrypt` to run on the threadpool in order to prevent blocking of the event loop. 

This fixes #2218 

The syntax is as follows:
```swift
req.password.async.verify()
```

### CODE REVIEW

**Comment by mcdappdev** [comment]
I'm a big fan of this - but is there any way to give those parameters defaults?

File: Sources/Vapor/Passwords/AsyncPasswordVerifier.swift [hunk 1]
```diff
@@ -0,0 +1,55 @@
+public struct AsyncPasswordVerifier {
+    private let verifier: PasswordVerifier
+    private let threadPool: NIOThreadPool
+    private let eventLoop: EventLoop
+    
+    public init (verifier: PasswordVerifier, threadPool: NIOThreadPool, eventLoop: EventLoop) {
```
Comment (tanner0101) [inline Sources/Vapor/Passwords/AsyncPasswordVerifier.swift:6]: nit: spacing

File: Sources/Vapor/Passwords/PasswordVerifier.swift [hunk 2]
```diff
@@ -20,4 +20,9 @@ extension PasswordVerifier {
             created: [UInt8](digest.utf8)
         )
     }
+    
+    public func asynchronized(on threadPool: NIOThreadPool, hopTo eventLoop: EventLoop) -> AsyncPasswordVerifier
```
Comment (tanner0101) [inline Sources/Vapor/Passwords/PasswordVerifier.swift:24]: I'd just call this `async`.

File: Sources/Vapor/Passwords/Request+Password.swift [hunk 3]
```diff
@@ -1,5 +1,31 @@
...
+        init(for req: Request) {
+            self.application = req.application
+            self.eventLoop = req.eventLoop
+        }
+        
+        public var async: AsyncPasswordVerifier {
+            self.application.password.asynchronized(on: application.threadPool, hopTo: eventLoop)
+        }
+        
+        public var verifier: PasswordVerifier {
```
Comment (tanner0101) [inline Sources/Vapor/Passwords/Request+Password.swift:19]: Maybe call this `sync` so the difference between the two is clear?

File: Sources/Vapor/Passwords/Request+Password.swift [hunk 4]
```diff
@@ -1,5 +1,31 @@
...
+            self.application.password
+        }
+        
+        public func verify<Password, Digest>(_ password: Password, created digest: Digest) throws -> Bool where Password : DataProtocol, Digest : DataProtocol {
```
Comment (tanner0101) [inline Sources/Vapor/Passwords/Request+Password.swift:23]: Indentation here would be nice.

File: Tests/VaporTests/BcryptTests.swift [hunk 5]
```diff
@@ -51,6 +51,30 @@ final class BcryptTests: XCTestCase {
...
+            .wait()
+        XCTAssertTrue(try BCryptDigest().verify("vapor", created: asyncHash))
+        
+        struct Verify: Content {
+            let digest: String
+        }
+        
+        app.post("test") { req -> EventLoopFuture<String> in
+            let verify = try req.content.decode(Verify.self)
+            return try req.password
```
Comment (tanner0101) [inline Tests/VaporTests/BcryptTests.swift:67]: 👍

File: Tests/VaporTests/BcryptTests.swift [hunk 6]
```diff
@@ -51,6 +51,30 @@ final class BcryptTests: XCTestCase {
         
         let result = try app.password.verify("vapor", created: hash)
         XCTAssertTrue(result)
+        
+        let asyncHash = try app.password
```
Comment (tanner0101) [inline Tests/VaporTests/BcryptTests.swift:55]: This + the next `app.post` test could probably go in separate test cases to keep things tidy.

File: Sources/Vapor/Passwords/AsyncPasswordVerifier.swift [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        _ password: Password,
+        created digest: Digest
+    ) throws -> EventLoopFuture<Bool>
+        where Password: DataProtocol, Digest: DataProtocol
+    {
+        let promise = eventLoop.makePromise(of: Bool.self)
+        self.threadPool.submit { _ in
+            do {
+                return promise.succeed(try self.verifier.verify(password, created: digest))
+            } catch  {
```
Comment (MrMage) [inline Sources/Vapor/Passwords/AsyncPasswordVerifier.swift:36]: drive-by nit (sorry): double space

File: Tests/VaporTests/PasswordTests.swift [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    
+    func testAsyncPlaintextApplicationPassword() throws {
+        let test = Environment(name: "testing", arguments: ["vapor"])
+        let app = Application(test)
+        defer { app.shutdown() }
+        
+        try assertAsyncApplicationPasswordVerifies(.plaintext, on: app)
+    }
+
```
Comment (0xTim) [inline Tests/VaporTests/PasswordTests.swift:60]: Can you add one final test to ensure the test works 😆 Basically just to make sure the async password uses the requested provider
```suggestion
    
    func testAsyncUsesProvider() throws {
         let test = Environment(name: "testing", arguments: ["vapor"])
         let app = Application(test)
         defer { app.shutdown() }
         app.passwords.use(.plaintext)
         let hash = try app.password.async(on: app.threadPool, hopTo: app.eventLoopGroup.next()).digest("vapor").wait()
         XCTAssertEqual(hash, "vapor")
     }
     
```
